akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -471,8 +483,29 @@ private[projection] class R2dbcOffsetStore(
 
       val offsetInserts = dao.insertTimestampOffsetInTx(conn, filteredRecords)
 
-      offsetInserts.map { _ =>
-        if (state.compareAndSet(oldState, evictedNewState))
+      val adoptable =
+        if (oldState.foreignLatestSortedByTimestamp.isEmpty || filteredRecords.isEmpty) Vector.empty
+        else {
+          val latestTimestamp = filteredRecords.maxBy(_.timestamp).timestamp
```

Comment:
This is using the records after deduplication. It would probably be better if it could trigger adoption as the projection moves forward in any way, even when replaying over events (because of skew between projection instances after scaling) but would need to then track in some way through the `validate` calls, where events will probably be deduplicated. Maybe it should keep track of the "current" timestamp in some way (different from the offset window or latest).

Reply:
Updated to a separate scheduled task, tracking the latest seen timestamp for both saved offsets and validate calls that return duplicates. https://github.com/akka/akka-projection/pull/1218/commits/2f13b7abe95a3b6197784ced4facaad6f1e88bc4